i6_models/parts/factored_hybrid/diphone.py

```diff
@@ -0,0 +1,167 @@
+            num_output=get_center_dim(cfg.n_contexts, cfg.num_hmm_states_per_phone, cfg.phoneme_state_class),
+            hidden_dim=cfg.context_mix_mlp_dim,
+            num_layers=cfg.context_mix_mlp_num_layers,
+            dropout=cfg.dropout,
+            activation=cfg.activation,
+        )
+
+    def forward(
+        self,
+        features: Tensor,  # B, T, F
+        contexts_left: Optional[Tensor] = None,  # B, T
+    ) -> Union[DiphoneLogitsV1, DiphoneProbsV1]:
+        """
+        :param features: Main encoder output. shape B, T, F. F=num_inputs
+        :param contexts_left: The left contexts used to compute p(c|l,x).
+            Shape during training: B, T. Must be `None` when forwarding as the model
+            will compute a flat, joint output scoring all possible contexts in forwarding
+            mode.
+        :return: During training: logits for p(c|l,x) and p(l|x).
+            During inference: *log* probs for p(c,l|x) in one large layer.
+        """
+
+        left_logits = self.left_context_encoder(features)  # B, T, C
+
+        if self.training:
+            assert contexts_left is not None
+            assert contexts_left.ndim >= 2
+        else:
+            assert contexts_left is None, "in eval mode, all left contexts are forwarded at the same time"
+            contexts_left = torch.arange(self.n_contexts)
+
+        # train: B, T, E
+        # eval: C, E
+        embedded_left_contexts = self.left_context_embedding(contexts_left)
+
+        if self.training:
+            # in training we forward exactly one context per T
+
+            center_features = torch.cat((features, embedded_left_contexts), -1)  # B, T, F+E
+            center_logits = self.center_encoder(center_features)  # B, T, C
+
+            return DiphoneLogitsV1(
+                embeddings_left_context=embedded_left_contexts, output_left=left_logits, output_center=center_logits
+            )
+        else:
+            # here we forward every context to compute p(c, l|x) = p(c|l, x) * p(l|x)
+
+            features = features.expand((self.n_contexts, -1, -1, -1))  # C, B, T, F
+            embedded_left_contexts_ = embedded_left_contexts.reshape((self.n_contexts, 1, 1, -1)).expand(
+                (-1, *features.shape[1:3], -1)
+            )  # C, B, T, E
+            center_features = torch.cat((features, embedded_left_contexts_), -1)  # C, B, T, F+E
+            center_logits = self.center_encoder(center_features)  # C, B, T, F'
+            center_probs = F.log_softmax(center_logits, -1)
+            center_probs = center_probs.permute((1, 2, 3, 0))  # B, T, F', C
+            left_probs = F.log_softmax(left_logits, -1)
+            left_probs = left_probs.unsqueeze(-2)  # B, T, 1, C
+
+            joint_probs = center_probs + left_probs  # B, T, F', C
+            joint_probs = torch.flatten(joint_probs, start_dim=2)  # B, T, joint
+
+            return DiphoneProbsV1(
```

Comment:
In the full diphone case we do not need the `output_left` or the embeddings, do we?

Would it make sense to use a different Dataclass here?

Reply:
We typically keep separate outputs for training and combine the two outputs only during inference, meaning that we do need a loss computation also on the `left_output`

Reply:
this is an argument in favor of having different kinds of outputs depending on train yes/no i.e. using different dataclasses (or e.g. tuples if we want to follow @JackTemaki s suggestion)

Reply:
Another aspect to consider @NeoLegends  is whether we want to have two separate modules for the architecture that uses frame-wise cross-entropy and the one that uses sequence-level cross-entropy, i.e. fullsum. Since in the latter also training would rely on the combination of the outputs.

Reply:
> sequence-level cross-entropy, i.e. fullsum.

*"e.g. fullsum"

Daniel has already offered to do a pr for the LFMMI op xP

Reply:
> how easy it is to change model classes from training to inference in torch

Let's also ask @SimBe195 and @Judyxujj, they are also running torch trainings for a while now

Reply:
I would not do it in separate modules, so that you always load everything consistently, but you can also not use `self.training`. What you can have is separate functions. `self.training` has the problem that it is really for "training" only, not e.g. for cross validation during training. We had bugs in setups before where we then applied dropout and specaugment for the validation sets.

So when you use `train_step` or `forward_step / onnx_export` you can call two different functions